Adding binding and library functions for setting Jacobian FD step size

PrasadBabarendaGamage opened the pull request:
Addresses #159 

Work in progress

PrasadBabarendaGamage committed:
Adding length dependent active contraction

PrasadBabarendaGamage committed:
Fixes for dataprojections

Distance vector sign was updated in previous commits by setting:
distanceVector(1:numberOfCoordinates)=interpolatedPoint%values(:,NO_PART_DERIV)-dataPointLocation
and by multiplying gradient and Hessian values by -1 compared with the original implementation. 

However, with the updated code, the gradient values are the same as what the old code produced, however the Hessian isn't. Fix was to revert to the original implementation for the gradient and Hessian and setting the distance vector as:
distanceVector(1:numberOfCoordinates)=dataPointLocation-interpolatedPoint%values(:,NO_PART_DERIV) 
This has been done for all cases of the projection (line, face, element).

PrasadBabarendaGamage committed:
Converting 1st PK active stress to 2 PK

PrasadBabarendaGamage committed:
Merge branch '149_projection_bug' into active_contraction

PrasadBabarendaGamage committed:
Adding element based active contraction parameter interpolation

Currently only for the Guccione constitutive relation

PrasadBabarendaGamage committed:
Adding binding and library functions for setting Jacobian FD step size

chrispbradley commented on src/opencmiss_iron.f90:
```diff
@@ -25199,6 +25201,38 @@ SUBROUTINE cmfe_Equations_JacobianMatricesTypesSet(equations,jacobianTypes,err)
+  !
+
+  !>Sets/changes the finite difference step size used for calculating the Jacobian
+  SUBROUTINE cmfe_Equations_JacobianFiniteDifferenceStepSizeSet(equations,jacobianFiniteDifferenceStepSize,err)
```
This has a strange interface? It is quite different from the Equations_JacobianGet interface? I would think that to specify a Jacobian to set the finite difference step size it should be the same as getting the Jacobian? There are also multiple step sizes passed via an array?

PrasadBabarendaGamage replied:
Hi Chris, yup I was very confused about this and why multiple step sizes were needed - I just used a similar function as JacobianMatricesTypesSet - which doesn't work anymore (please see https://github.com/OpenCMISS/iron/issues/160), but will try to use Equations_JacobianGet and hopefully this fixes the issue.

PrasadBabarendaGamage replied:
Hi @chrispbradley I don't think Equations_JacobianGet will do the job as that is called after Equations are finished, whereas the step size needs to be set during setting up of Equations, Problem is that both cmfe_Equations_JacobianFiniteDifferenceStepSizeSet and JacobianMatricesTypesSet won't work during Equations setup as they require vectorMatrices which I don't think are setup until equations are finished (#160 ) - most likely need to be cached. Unfortunately, Mario can't move forward without these and it has already taken too long to try and set a simple flag. I am going to place the stepsize and Jacobian type set directly in the equations type and get it going so that he can move forward with debugging the active contraction (still doesn't work). Then can get back to fixing this setup issue. Thanks

PrasadBabarendaGamage replied:
Once fixed, will need to add some more cases to the finite elasticity tests to make sure future code changes that effect it are flagged.

PrasadBabarendaGamage replied:
Added more info to https://github.com/OpenCMISS/iron/issues/160 regarding the source of the issue

chrispbradley replied:
Hi Prasad, sure, add it to the equations. Maybe make the routine name Equations_FiniteDifferenceStepSizeSet since this will apply to the equations rather than just Jacobians? I'm also not sure why JacobianVariableTypesSet is a public routine???

PrasadBabarendaGamage replied:
Thanks, will do. The JacobianVariableTypesSet was added by Jenny to allow a switch between calculating analytic and FD Jacobians inside OpenCMISS. These options only work if nonlinearSolver.NewtonJacobianCalculationTypeSet(
    iron.JacobianCalculationTypes.EQUATIONS) 
If nonlinearSolver.NewtonJacobianCalculationTypeSet(
    iron.JacobianCalculationTypes.FD) then the petsc's FD calculates the Jacobian

chrispbradley replied:
Hmm, but the variableType for the Jacobian is the field variable the residual is differentiated by to give the Jacobian. This is something that is defined by the variable types for the equations. I'm not sure I get why a user would want or need to change them?

PrasadBabarendaGamage replied:
yup, thats a bit weird, we probably need to update the name to equations.JacobianCalculationType?

chrispbradley commented on src/equations_matrices_routines.f90:
```diff
@@ -3005,6 +3008,60 @@ END SUBROUTINE EquationsMatrices_JacobianTypesSet
+
+    !Argument variables
+    TYPE(EquationsMatricesVectorType), POINTER :: vectorMatrices !<A pointer to the vector equations matrices to set the Jacobian finite difference step sizes for.
+    REAL(DP), INTENT(IN) :: jacobianStepSizes(:) !<jacobianStepSizes(matrixIdx). The Jacobian finite difference step sizes for the matrixIdx'th Jacobian matrix.
```
Not sure why multiple step sizes are allowed like this as there is no information on what Jacobian the step size is for? If there were multiple Jacobians and you only specified one step size then what Jacobian are you trying to set the Jacobian for?

PrasadBabarendaGamage committed:
Temporarily fixing issues with setting Jacobian type and fd step size

For debugging active contraction. Some changes will need to be reverted when these attributes are correctly set.

PrasadBabarendaGamage committed:
Merge branch 'active_contraction' into 159_temp_jacobian_type_fd_step_fix

PrasadBabarendaGamage committed:
Generalising finite elasticity analytic jacobian 

Now accepts different independent field interpolation.
Field constant value has been hard coded to 0.1

PrasadBabarendaGamage committed:
Merge branch '159_temp_jacobian_type_fd_step_fix' into 159_fd_step_size

PrasadBabarendaGamage commented on src/equations_matrices_routines.f90:
```diff
@@ -2956,46 +2956,24 @@ END SUBROUTINE EquationsMatrices_JacobianOutput
-        CALL FlagError(localError,err,error,*999)
-      END SELECT
-    ENDDO !matrixIdx
+    IF(.NOT.ASSOCIATED(equations)) CALL FlagError("Equations is not associated.",err,error,*999)
```
creating link

chrispbradley commented:
OK, looking at this again as part of the pull request I think the best way to resolve this is not have the Jacobian type and step size in the equations as this will apply to all Jacobians. I think the answer is to not set the jacobian and step size inside the equations create start/finish block. We don't even have any equations mappings let alone matrices until after equations create finish is called. We thus don't even know if we have any Jacobians then. The Equations_JacobianMatrixGet is used after equations create finish and so I think the Jacobian calcuation type and step size can be set then? i.e., after equations create finish?
